src/stream/src/executor/sync_kv_log_store.rs

```diff
@@ -0,0 +1,1045 @@
+                Self::try_next_flushed_chunk_future(read_flushed_chunk_future).await
+            }
+            LogStoreBufferItem::Barrier { next_epoch, .. } => {
+                // FIXME(kwannoel): Is `next_epoch` correct for truncation??
+                truncation_offset.replace((next_epoch, None));
+                Ok(None)
+            }
+            LogStoreBufferItem::UpdateVnodes(_) => Ok(None),
+        }
+    }
+}
+
+// Write methods
+impl<S: StateStore> SyncedKvLogStoreExecutor<S> {
+    #[allow(clippy::too_many_arguments)]
+    async fn write_barrier(
+        state_store: &mut S::Local,
+        serde: &mut LogStoreRowSerde,
+        barrier: Barrier,
+        metrics: &mut KvLogStoreMetrics,
+        truncation_offset: Option<ReaderTruncationOffsetType>,
+        seq_id: &mut SeqIdType,
+        buffer: &mut SyncedLogStoreBuffer,
+        actor_id: ActorId,
+    ) -> StreamExecutorResult<()> {
+        let epoch = state_store.epoch();
+        let mut flush_info = FlushInfo::new();
+
+        // FIXME(kwannoel): Handle paused stream.
+        for vnode in serde.vnodes().iter_vnodes() {
+            let (key, value) = serde.serialize_barrier(epoch, vnode, barrier.is_checkpoint());
+            flush_info.flush_one(key.estimated_size() + value.estimated_size());
+            state_store.insert(key, value, None)?;
+        }
+
+        // FIXME(kwannoel): Flush all unflushed chunks
+        // As an optimization we can also change it into flushed items instead.
+        // This will reduce memory consumption of logstore.
+
+        flush_info.report(metrics);
+
+        // Apply truncation
+        let watermark = truncation_offset.map(|truncation_offset| {
+            VnodeWatermark::new(
+                serde.vnodes().clone(),
+                serde.serialize_truncation_offset_watermark(truncation_offset),
+            )
+        });
+        state_store.flush().await?;
+        let watermark = watermark.into_iter().collect_vec();
+        state_store.seal_current_epoch(
+            barrier.epoch.curr,
+            SealCurrentEpochOptions {
+                table_watermarks: Some((WatermarkDirection::Ascending, watermark)),
+                switch_op_consistency_level: None,
+            },
+        );
+
+        // Add to buffer
+        buffer.buffer.push_back((
+            epoch,
+            LogStoreBufferItem::Barrier {
+                is_checkpoint: barrier.is_checkpoint(),
+                next_epoch: barrier.epoch.curr,
+            },
+        ));
+        buffer.next_chunk_id = 0;
+        buffer.update_unconsumed_buffer_metrics();
+
+        // Apply Vnode Update
+        if let Some(vnode_bitmap) = barrier.as_update_vnode_bitmap(actor_id) {
+            state_store.update_vnode_bitmap(vnode_bitmap.clone());
+            serde.update_vnode_bitmap(vnode_bitmap.clone());
+            buffer
+                .buffer
+                .push_back((epoch, LogStoreBufferItem::UpdateVnodes(vnode_bitmap)));
```

Comment:
We don't need to push this `UpdateVnodes` item. And in other code appearance we shall mark it as unreachable.